tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java

```diff
@@ -0,0 +1,261 @@
+ * and ExternalCommandWorker should follow these APIs:
+ *
+ * Control API: {"action":"stop"|"start", "spec":ExternalCommandSpec}
+ *
+ * ExternalCommandWorker can control the external process by sending commands to the process's stdin.
+ *
+ * Start a task {"action":"start", "spec":ExternalCommandSpec}.
+ *
+ * Stop a task {"action":"stop", "spec":ExternalCommandSpec}.
+ *
+ *
+ * Communication API: {"status":val, "log":val, "error":"msg"}
+ *
+ * The external process can sends JSON object messages to the ExternalCommandWorker via the process's stduot, and error
+ * messages via the process's stderr.
+ * ExternalCommandWorker should continuously monitor the stdout and the stderr of the external process line by line.
+ *
+ * For any JSON object the external process writes to its stdout, if the object has the "status" field, this worker should set
+ * its status with the "status" value, if the object has the "error" field and the value is a JSON string, this worker should
+ * set the agent error message with the "error" value, and if the object has the "log" field, this worker should put the "log"
+ * value it its log.
+ *
+ * For any line the external process writes to its stderr, this worker should save the line to its error log and set the
+ * error message with the line.
+ */
+public class ExternalCommandWorker implements TaskWorker {
+    private static final Logger log = LoggerFactory.getLogger(ExternalCommandWorker.class);
+
+    private final String id;
+
+    private final ExternalCommandSpec spec;
+
+    private final AtomicBoolean running = new AtomicBoolean(false);
+
+    private ScheduledExecutorService executor;
+
+    private Process process;
+
+    private PrintWriter controlChannel;
+
+    private String errMsg = "";
+
+    private WorkerStatusTracker status;
+
+    private KafkaFutureImpl<String> doneFuture;
+
+    public ExternalCommandWorker(String id, ExternalCommandSpec spec) {
+        this.id = id;
+        this.spec = spec;
+    }
+
+    @Override
+    public void start(Platform platform, WorkerStatusTracker status,
+                      KafkaFutureImpl<String> doneFuture) throws Exception {
+        if (!running.compareAndSet(false, true)) {
+            throw new IllegalStateException("ExternalCommandWorker is already running.");
+        }
+        log.info("{}: Activating ExternalCommandWorker with {}", id, spec);
+        this.executor = Executors.newScheduledThreadPool(3, // processMonitor, updater and errorUpdater
+            ThreadUtils.createThreadFactory("ExternalCommandWorkerThread%d", false));
+        this.status = status;
+        this.doneFuture = doneFuture;
+        executor.submit(new ProcessMonitor());
+    }
+
+    public class ProcessMonitor implements Runnable {
+        @Override
+        public void run() {
+            try {
+                if (spec.command().isEmpty()) {
+                    errMsg = "Empty Command";
```

Comment:
Maybe something like "No command specified" would be more descriptive

Reply:
Done.